django/views/generic/list.py

```diff
@@ -181,7 +181,13 @@ def get_template_names(self):
         if hasattr(self.object_list, 'model'):
             opts = self.object_list.model._meta
             names.append("%s/%s%s.html" % (opts.app_label, opts.model_name, self.template_name_suffix))
-
+        else:
+            if self.object_list is None and not names:
```

Comment:
Collapse to `elif`

Reply:
Also check whether this is an instance of `QuerySet` rather than just `None` if it must be a `QuerySet`?

Reply:
I'll have to double check this. Turns out get_queryset may also return a list (or other iterable) - but I guess if it's not a QuerySet and no template names are found... it would make sense to raise ImproperlyConfigured rather than TypeError, but would also need to adapt the error message.